src/hotspot/share/gc/parallel/parallelScavengeHeap.hpp

```diff
@@ -288,4 +292,30 @@ class AdaptiveSizePolicyOutput : AllStatic {
   }
 };
 
+// The HeapBlockClaimer is used during parallel iteration over heap,
+// allowing workers to claim heap blocks, gaining exclusive rights to these blocks.
+// The eden, survivor spaces are treated as single blocks as it is hard to divide
+// these spaces.
+// The old spaces are divided into serveral fixed-size blocks.
+class HeapBlockClaimer : public StackObj {
+  uint           _n_workers;
+  uint           _n_blocks;
+  uint           _unclaimed_index;
+  volatile uint* _claims;
+
+  static const uint Unclaimed = 0;
+  static const uint Claimed   = 1;
+
+  public:
+  HeapBlockClaimer(uint n_workers);
+  ~HeapBlockClaimer();
+
+  // Claim the block and get the block index.
+  bool claim_and_get_block(uint* block_index);
+
+  static const uint eden_index = 0;
+  static const uint survivor_index = 1;
+  static const uint num_inseparable_spaces = 2;
```

Comment:
One more comment. These should be CamelCase, and I have an idea to get rid of `num_inseparable_spaces`. 

If we make `_claimed_index` an integral type we could define `EdenIndex = -2` and `SurvivorIndex = -1`. If we then initialize `_claimed_index = EdenIndex` the old gen indices will be correct without needing a decrease.

Reply:
Good idea! I also suggest to rename _n_blocks to _max_iteratable_block_index in HeapBlockClaimer, To avoid incorrectly considering _n_blocks to be "total number of blocks to be iterated".

Reply:
Or the _n_blocks could be deleted,  I will update the PR soon.